packages/core/src/render3/reactivity/linked_signal.ts

```diff
@@ -140,95 +57,33 @@ export function linkedSignal<S, D>(
-  node.source = isShorthand ? optionsOrComputation : optionsOrComputation.source;
-  if (!isShorthand) {
-    node.computation = optionsOrComputation.computation as ComputationFn<unknown, unknown>;
+  if (typeof optionsOrComputation === 'function') {
+    const getter = createLinkedSignal<D, D>(
+      optionsOrComputation,
+      identityFn<D>,
+      options?.equal,
+    ) as LinkedSignalGetter<D, D> & WritableSignal<D>;
+    return getterAsWritableSignal(getter);
+  } else {
+    const getter = createLinkedSignal<S, D>(
+      optionsOrComputation.source,
+      optionsOrComputation.computation,
+      optionsOrComputation.equal,
+    ) as LinkedSignalGetter<S, D> & WritableSignal<D>;
+    return getterAsWritableSignal(getter);
   }
-  const equal = isShorthand ? options?.equal : optionsOrComputation.equal;
-  if (equal) {
-    node.equal = equal as ValueEqualityFn<unknown>;
-  }
-  return createLinkedSignal(node as LinkedSignalNode<S, D>);
 }
 
-/**
- * A dedicated symbol used before a state value has been set / calculated for the first time.
- * Explicitly typed as `any` so we can use it as signal's value.
- */
-const UNSET: any = /* @__PURE__ */ Symbol('UNSET');
-
-/**
- * A dedicated symbol used in place of a linked signal value to indicate that a given computation
- * is in progress. Used to detect cycles in computation chains.
- * Explicitly typed as `any` so we can use it as signal's value.
- */
-const COMPUTING: any = /* @__PURE__ */ Symbol('COMPUTING');
-
-/**
- * A dedicated symbol used in place of a linked signal value to indicate that a given computation
- * failed. The thrown error is cached until the computation gets dirty again or the state is set.
- * Explicitly typed as `any` so we can use it as signal's value.
- */
-const ERRORED: any = /* @__PURE__ */ Symbol('ERRORED');
-
-// Note: Using an IIFE here to ensure that the spread assignment is not considered
-// a side-effect, ending up preserving `LINKED_SIGNAL_NODE` and `REACTIVE_NODE`.
-// TODO: remove when https://github.com/evanw/esbuild/issues/3392 is resolved.
-const LINKED_SIGNAL_NODE = /* @__PURE__ */ (() => {
-  return {
-    ...REACTIVE_NODE,
-    value: UNSET,
-    dirty: true,
-    error: null,
-    equal: defaultEquals,
-    computation: identityFn,
-
-    producerMustRecompute(node: LinkedSignalNode<unknown, unknown>): boolean {
-      // Force a recomputation if there's no current value, or if the current value is in the
-      // process of being calculated (which should throw an error).
-      return node.value === UNSET || node.value === COMPUTING;
-    },
-
-    producerRecomputeValue(node: LinkedSignalNode<unknown, unknown>): void {
-      if (node.value === COMPUTING) {
-        // Our computation somehow led to a cyclic read of itself.
-        throw new Error('Detected cycle in computations.');
-      }
+function getterAsWritableSignal<S, D>(getter: LinkedSignalGetter<S, D> & WritableSignal<D>) {
```

Comment:
Minor naming and typing nit.

I was initially a bit confused by this function's name as it sounds read-only, but in fact it modifies its argument.

Another small improvement is to not require the caller to cast to `WritableSignal<D>` already, but to return the casted argument.

```suggestion
function upgradeLinkedSignalGetter<S, D>(getter: LinkedSignalGetter<S, D>): WritableSignal<D> {
```